vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.h

```diff
@@ -0,0 +1,119 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#pragma once
+
+#include "atomic_entry_ref.h"
+#include <vespa/vespalib/util/array.h>
+#include <vespa/vespalib/util/arrayref.h>
+#include <vespa/vespalib/util/generationhandler.h>
+#include <limits>
+#include <atomic>
+#include <deque>
+#include <functional>
+
+namespace vespalib { class GenerationHolder; }
+namespace vespalib::datastore {
+
+class EntryComparator;
+
+/*
+ * Fixed sized hash map over keys in data store, meant to support a faster
+ * dictionary for unique store with relation to lookups.
+ *
+ * Currently hardcoded key and data types, where key references an entry
+ * in a UniqueStore and value references a posting list
+ * (cf. search::attribute::PostingStore).
+ *
+ * This structure supports one writer and many readers.
+ *
+ * A reader must own an appropriate GenerationHandler::Guard to ensure
+ * that memory is held while it can be accessed by reader.
+ *
+ * The writer must update generation and call transfer_hold_lists and
+ * trim_hold_lists as needed to free up memory no longer needed by any
+ * readers.
+ */
+class FixedSizeHashMap {
+public:
+    static constexpr uint32_t no_node_idx = std::numeric_limits<uint32_t>::max();
+    using KvType = std::pair<AtomicEntryRef, AtomicEntryRef>;
+    using generation_t = GenerationHandler::generation_t;
+    using sgeneration_t = GenerationHandler::sgeneration_t;
+
+private:
+    class ChainHead {
+        std::atomic<uint32_t> _node_idx;
+
+    public:
+        ChainHead()
+            : _node_idx(no_node_idx)
+        {
+        }
+        // Writer thread
+        uint32_t load_relaxed() const noexcept { return _node_idx.load(std::memory_order_relaxed); }
+        void set(uint32_t node_idx) { _node_idx.store(node_idx, std::memory_order_release); }
+
+        // Reader thread
+        uint32_t load_acquire() const noexcept { return _node_idx.load(std::memory_order_acquire); }
+    };
+    class Node {
+        KvType _kv;
+        std::atomic<uint32_t> _next;
+    public:
+        Node()
+            : Node(std::make_pair(AtomicEntryRef(), AtomicEntryRef()), no_node_idx)
+        {
+        }
+        Node(KvType kv, uint32_t next)
+            : _kv(kv),
+              _next(next)
+        {
+        }
+        Node(Node &&rhs); // Must be defined, but must never be used.
+        void on_free();
+        std::atomic<uint32_t>& get_next() noexcept { return _next; }
+        const std::atomic<uint32_t>& get_next() const noexcept { return _next; }
+        KvType& get_kv() noexcept { return _kv; }
+        const KvType& get_kv() const noexcept { return _kv; }
+    };
+
+    Array<ChainHead>  _chain_heads;
```

Comment:
I wonder if it is better to use std::vector<T, vespalib::allocator_large<T>> instead of Array. It should give the same advantages on allocation, and supports debugging better and is also more standard.
Ideally we should have something even simpler that does not require move or copy constructors.